src/notification/verifyIosNotificationData.js

```diff
@@ -0,0 +1,232 @@
+
+
+    This has been expanded incrementally over the years. As of 2020-02, commit
+    2.2-dev-775-g10e7e15088, the current form of APNs messages is as follows:
+
+    ```
+    type StreamData = {
+        // added 1.7.0-1351-g98943a8333, release 1.8.0+
+        recipient_type: 'stream',
+        stream: string,
+        topic: string,
+    };
+
+    type PmData = {
+        // added 1.7.0-1351-g98943a8333, release 1.8.0+
+        recipient_type: 'private',
+
+        // added 1.7.0-2360-g693a9a5e70, release 1.8.0+
+        // present only on huddles (multi-user PMs)
+        pm_users?: string,  // CSV of int (user ids)
+    };
+
+    type Data = { zulip: {
+        // present since antiquity
+        message_ids: [number],  // single-element tuple!
+
+        // added 1.7.0-1351-g98943a8333, release 1.8.0+
+        sender_email: string,
+        sender_id: number,
+        server: string,     // settings.EXTERNAL_HOST
+        realm_id: number,   // server-internal realm identifier
+
+        // added 1.8.0-2150-g5f8d193bb7, release 1.9.0+
+        realm_uri: string,  // as `/server_settings` response
+
+        // added 2.1-dev-540-g447a517e6f, release 2.1.0+
+        user_id: number,    // recipient id
+
+        ...(StreamData | PmData),
+    } };
+    ```
+
+    Note that prior to 1.7.0-1351-g98943a8333, we only received the
+    `message_ids` field. Messages of this form are not useful.
+
+    The pair (server, realm_id) uniquely identifies a Zulip realm, and was
+    originally intended to permit the client to associate a notification with
+    its associated realm. Unfortunately, there is no way to get either of these
+    from a server via the API.
+*/
+
+/** Local error type. */
+class ApnsMsgValidationError extends Error {
+  extras: JSONable;
+  constructor(message, extras: JSONable) {
+    super(message);
+    this.extras = extras;
+  }
+}
+
+const RECIPIENT_TYPES = Object.freeze({ stream: true, private: true });
+
+/** Private. Exported only for tests. */
+//
+// Extract Zulip notification data from a JSONable dictionary imported from an
+// APNs notification.
+//
+// @returns A `Notification` on success, `undefined` on suppressible failure.
+// @throws An ApnsMsgValidationError on interesting failure.
+//
+export const impl = (rawData: JSONableDict): Notification | void => {
+  /** Helper function: fail. */
+  const err = (style: string) =>
+    new ApnsMsgValidationError(`Received ${style} APNs notification`, { data: rawData });
+
+  // APNs messages are JSON dictionaries. The `aps` entry of this dictionary is
+  // required, with a structure defined by Apple; all other entries are
+  // available to the application.
+  //
+  // The react-native-notifications library filters out `aps`, parses it, and
+  // hands us the rest as "data". Pretty much any iOS notifications library
+  // should do the same, but we don't rely on that.
+
+  const data: PartialJSONableDict = (() => {
```

Comment:
To prevent accidentally hitting the indexer-property soundness hole. Flow is indeed happy if you do that, but Flow will remain happy if you do that and then make a lot of accesses to `data` that won't work.

(This could stand to be in a comment.)

Reply:
Could you make a version of `PartialJSONableDict` that allows for `void`s to appear recursively? (Perhaps in `jsonable.js` so it's next to the definition of `JSONable`, with which it shares some facts.) Then it should be possible to mark just one thing at the outside as "partial", and Flow would catch if we dropped the partialness somewhere along the way.

That one spot at the outside would then be a good spot for a short comment mentioning why it's needed.

Reply:
> That one spot at the outside would then be a good spot for a short comment mentioning why it's needed.

That deserves a long comment, really. (Done.)